subcommand/consul-sidecar/command.go

```diff
@@ -80,57 +96,165 @@ func (c *Command) Run(args []string) int {
 	for {
 		start := time.Now()
-		cmd := exec.CommandContext(ctx, c.flagConsulBinary, c.consulCommand...)
+		cmd := exec.CommandContext(signalCtx, c.flagConsulBinary, c.consulCommand...)
 
 		// Run the command and record the stdout and stderr output
 		output, err := cmd.CombinedOutput()
 		if err != nil {
-			logger.Error("failed to sync service", "output", strings.TrimSpace(string(output)), "err", err, "duration", time.Since(start))
+			c.logger.Error("failed to sync service", "output", strings.TrimSpace(string(output)), "err", err, "duration", time.Since(start))
 		} else {
-			logger.Info("successfully synced service", "output", strings.TrimSpace(string(output)), "duration", time.Since(start))
+			c.logger.Info("successfully synced service", "output", strings.TrimSpace(string(output)), "duration", time.Since(start))
 		}
 		select {
 		// Re-loop after syncPeriod or exit if we receive interrupt or terminate signals.
 		case <-time.After(c.flagSyncPeriod):
 			continue
-		case <-ctx.Done():
+		case <-signalCtx.Done():
+			// After the signal is received, wait for the merged metrics server
+			// to gracefully shutdown as well if it has been enabled. This can
+			// take up to metricsServerShutdownTimeout seconds.
+			if c.flagEnableMetricsMerging {
+				c.logger.Info("Attempting to shut down metrics server.")
+				c.shutdownMetricsServer(server)
+			}
 			return 0
 		}
 	}
 }
 
+// shutdownMetricsServer handles gracefully shutting down the server. This will
+// call server.Shutdown(), which will indefinitely wait for connections to turn
+// idle. To avoid potentially waiting forever, we pass a context to
+// server.Shutdown() that will timeout in metricsServerShutdownTimeout (5) seconds.
+func (c *Command) shutdownMetricsServer(server *http.Server) {
+	shutdownCtx, shutdownCancelFunc := context.WithTimeout(context.Background(), metricsServerShutdownTimeout)
+	defer shutdownCancelFunc()
+
+	// If the server hasn't started up yet before this function is called, it
+	// could be nil.
+	if server != nil {
+		c.logger.Info("Merged metrics server exists, attempting to gracefully shut down server.")
+		if err := server.Shutdown(shutdownCtx); err != nil {
+			c.logger.Error(fmt.Sprintf("Server shutdown failed: %s", err))
+		}
+		c.logger.Info("Server has been shut down.")
+	}
+}
+
+// createMergedMetricsServer sets up the merged metrics server.
+func (c *Command) createMergedMetricsServer() *http.Server {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/stats/prometheus", c.mergedMetricsHandler)
+
+	mergedMetricsServerAddr := fmt.Sprintf("127.0.0.1:%s", c.flagMergedMetricsPort)
+	server := &http.Server{Addr: mergedMetricsServerAddr, Handler: mux}
+
+	return server
+}
+
+// mergedMetricsHandler has the logic to append both Envoy and service metrics
+// together, logging if it's unsuccessful at either.
+func (c *Command) mergedMetricsHandler(rw http.ResponseWriter, r *http.Request) {
```

Comment:
I was giving this workflow a read and had a general question/suggestion:
Does it make sense to read and write the `envoy` metrics and then attempt to read and then write the `app` metrics? It might simplify some of the error handling and plus you wouldnt have to check if the response from the application metrics end point was `nil` as you can determine that from if the read attempt returned an error.
Just a thought.

Reply:
Ah I like that flow. I changed it to match that, and added a return if there was an error trying to netClient.Get(appMetricsAddr), since we will have already written envoy metrics at that point.